### ISSUE TITLE
VMFPLC2 utility rewrite does not build on MacOS

### ISSUE THREAD

**Issue by Peter-J-Jansen** [SDL-Hercules-390/hyperion#188]
The recent work on VMFPLC2 unfortunately causes the SDL-Hyperion build to fail on MacOS. It ends like this : 

```
[…]
 CC       faketape.lo
  CC       hettape.lo
  CC       omatape.lo
  CC       scsitape.lo
  CC       scsiutil.lo
  CCLD     hdt3420.la

*** Warning: Linking the shared library hdt3420.la against the
*** static library ../SDL-extpkgs/lib/libcrypto64.a is not portable!

*** Warning: Linking the shared library hdt3420.la against the
*** static library ../SDL-extpkgs/lib/libdecNumber64.a is not portable!

*** Warning: Linking the shared library hdt3420.la against the
*** static library ../SDL-extpkgs/lib/libSoftFloat64.a is not portable!

*** Warning: Linking the shared library hdt3420.la against the
*** static library ../SDL-extpkgs/lib/libtelnet64.a is not portable!
  CCLD     vmfplc2

*** Warning: Linking the executable vmfplc2 against the loadable module
*** hdt3420.so is not portable!
** Warning, lib hdt3420.so is a module, not a shared library

** And there doesn't seem to be a static archive available
** The link will probably fail, sorry
ld: can't link with bundle (MH_BUNDLE) only dylibs (MH_DYLIB) file './.libs/hdt3420.so' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [vmfplc2] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
pjvmojave:SDL-hyperion hercules$ make check
Making check in m4
make[1]: Nothing to be done for `check'.
Making check in util
make[1]: Nothing to be done for `check'.
Making check in html
make[1]: Nothing to be done for `check'.
Making check in man
make[1]: Nothing to be done for `check'.
Making check in .
  CCLD     vmfplc2

*** Warning: Linking the executable vmfplc2 against the loadable module
*** hdt3420.so is not portable!
** Warning, lib hdt3420.so is a module, not a shared library

** And there doesn't seem to be a static archive available
** The link will probably fail, sorry
ld: can't link with bundle (MH_BUNDLE) only dylibs (MH_DYLIB) file './.libs/hdt3420.so' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [vmfplc2] Error 1
make: *** [check-recursive] Error 1
pjvmojave:SDL-hyperion hercules$ 
```

**Comment by Fish-Git** [comment]
I have no idea why this is happening.&nbsp;`:(`

It works fine on non-Mac, so it's something unique to the Apple Mac and I'm afraid I'm not skilled enough with Apple Mac to figure out what's going on or how to fix it.

I'm sorry, but this one is beyond my capabilities, Peter. Someone else is going to have to figure it out.&nbsp;`:(`

**Comment by joemonk64** [comment]
This is probably due to the fact that the -fPIC flag was not included. Shared libraries (ending in .a extension) on x86_64 MUST be processed with that flag.

**Comment by Fish-Git** [comment]
Good guess, but wrong.

All of the External Package .a _static_ libraries (not shared libraries) _were_ built with the `-fPIC` option:

https://github.com/SDL-Hercules-390/SoftFloat/blob/1c66591b529ecd8b6ab42f7c588604d67a9759da/CMakeLists.txt#L104-L110

The problem is not the:

```
*** Warning: Linking the shared library hdt3420.la against the
*** static library ../SDL-extpkgs/lib/libSoftFloat64.a is not portable!
```

warning messages, but rather is either:

```
*** Warning: Linking the executable vmfplc2 against the loadable module
*** hdt3420.so is not portable!
** Warning, lib hdt3420.so is a module, not a shared library
```

and/or:

```
ld: can't link with bundle (MH_BUNDLE) only dylibs (MH_DYLIB) file './.libs/hdt3420.so' for architecture x86_64
```

(although both of them could be describing the same error/problem).

I think it has something to do with the difference between ".dylib" vs. ".so" on the Mac, but I don't know enough about it to know how to fix it.

The change that seems is triggering the error is in `Makefile.am`:

https://github.com/SDL-Hercules-390/hyperion/blob/155c59747a6d20d4c2262742cc4c5f2deb5a93a4/Makefile.am#L680-L682

It works fine on on e.g. CentOS and Ubuntu, etc, but not on Apple Mac, and I don't know why!&nbsp;`:(`

**Comment by Fish-Git** [comment]
> Shared libraries (ending in .a extension) ...

Shared libraries don't end with a `.a` file extension.  They end with a `.so` file extension. _Static link_ libraries end with a `.a` file extension. (And as mentioned in my previous comment, _all_ of the .a static link libraries are built with -fPIC.

**Comment by joemonk64** [comment]
OK ... 

"Mach-O shared libraries have the file type MH_DYLIB and carry the extension .dylib. They can be linked against with the usual static linker flags, e.g. -lfoo for libfoo.dylib. They can be created by passing the -dynamiclib flag to the compiler. (-fPIC is the default and needn't be specified.)

Loadable modules are called "bundles" in Mach-O speak. They have the file type MH_BUNDLE. They can carry any extension; the extension .bundle is recommended by Apple, but most ported software uses .so for the sake of compatibility. Typically, you'll use bundles for plug-ins that extend an application; in such situations, the bundle will link against the application binary to gain access to the application’s exported API. They can be created by passing the -bundle flag to the compiler.

Both dylibs and bundles can be dynamically loaded using the dl APIs (e.g. dlopen, dlclose). It is not possible to link against bundles as if they were shared libraries. However, it is possible that a bundle is linked against real shared libraries; those will be loaded automatically when the bundle is loaded."

"On ELF systems like Linux, both use the same file format; any piece of shared code can be used as a library and for dynamic loading."

So, the linker thinks you are trying to link against a bundle, which is a no-no on a mac.

**Comment by joemonk64** [comment]
Are you using -output-obj or -dynamiclib when you compile? -output-obj produces bundles, -dynamiclib produces a shared library.

**Comment by Juergen-Git** [comment]
Hi Peter

as a temporary local circumvention you can edit Makefile and libtool to do the "right" thing. Modulo location in your files, it could look like this in Unix patch syntax:

diff -Naur a/Makefile b/Makefile
--- a/Makefile	2019-02-24 17:41:23.000000000 +0100
+++ b/Makefile	2019-02-24 13:51:27.000000000 +0100
@@ -1047,7 +1047,12 @@   Main Makefile only!!!
 hdt3270_la_LDFLAGS = $(DYNMOD_LD_FLAGS)
 hdt3270_la_LIBADD = $(DYNMOD_LD_ADD)
 hdt3420_la_SOURCES = tapedev.c tapeccws.c awstape.c faketape.c hettape.c omatape.c scsitape.c scsiutil.c
-hdt3420_la_LDFLAGS = $(DYNMOD_LD_FLAGS)
+hdt3420_la_LDFLAGS = \
+  -dynamiclib        \
+  -no-undefine d     \
+  $(XSTATIC)         \
+  -export-dynamic    \
+  -avoid-version
 hdt3420_la_LIBADD = $(DYNMOD_LD_ADD) libherct.la
 hdt3505_la_SOURCES = cardrdr.c sockdev.c
 hdt3505_la_LDFLAGS = $(DYNMOD_LD_FLAGS)

diff -Naur a/libtool b/libtool
--- a/libtool	2019-02-24 17:39:55.000000000 +0100
+++ b/libtool	2019-02-24 14:17:57.000000000 +0100
@@ -3104,9 +3104,9 @@
 	;;
       *)
 	if test "$module" = no; then
-	  $echo "$modename: libtool library \`$output' must begin with \`lib'" 1>&2
+	  $echo "$modename: libtool library \`$output' should begin with \`lib'" 1>&2
 	  $echo "$help" 1>&2
-	  exit 1
+	# exit 1
 	fi
 	if test "$need_lib_prefix" != no; then
 	  # Add the "lib" prefix for modules if required

libtool will still barf on the "wrong" name of hdt3420.dylib, but it will work..

Cheers
Jürgen

**Comment by Juergen-Git** [comment]
Hi Peter

as a temporary local circumvention you can edit Makefile and libtool to do the "right" thing. Modulo location in your files, it could look like this in Unix patch syntax:

--- Sorry, I don't seem to get the patch in here, no idea how to make Github simply take in text ---
--- see link below ---

libtool will still barf on the "wrong" name of hdt3420.dylib, but it will work...

Cheers
Jürgen

[patch.txt](https://github.com/SDL-Hercules-390/hyperion/files/2907430/patch.txt)

**Comment by Fish-Git** [comment]
_**OFF TOPIC**_

> Sorry, I don't seem to get the patch in here, no idea how to make Github simply take in text. See link below .

FYI: I've been experiencing "unable to render preview" errors with increasing frequency when trying to post comments on GitHub. I don't know if it's something Microsoft changed or why it's happening, but it's _extremely annoying_ when it happens.

It occured just now while trying to edit your comment, Jürgen, so as a test, I tried creating a simple test comment:

    This is a test.

    ```
    foo
    bar
    ```

    This is a test.

But even something as simple as that resulted in _"error rendering preview"._ (Sheesh!)

I clicked on "Contact GitHub" at the bottom of the page to compose a complaint (bug report), but because I forgot to do some screen grabs beforehand, I "duplicated" my browser tab (Chrome) so I could return to the original GitHub Issue web page (i.e. _THIS_ web page you are reading right now), and retried my simple test that _was_ failing (so I could grab a screenshot for my bug report), BUT LO AND BEHOLD _it started working this time!_ (wtf?)

So I don't know what I did to cause it to start working, but now I can preview my comments before I post them.

So I edited your comment, Jürgen, to display your patch inline in your comment. I trust that's okay.

**Comment by Fish-Git** [comment]
> ```diff
> +  -no-undefine d     \
> ```

@Juergen-Git Jürgen, is the above correct? It looks like a typo. Shouldn't the above actually be `-no-undefined` without the space?

**Comment by Juergen-Git** [comment]
Yes, it should of course read -no-undefined without the blank. I must have messed this up during the many copy&paste tries that didn‘t work...

**Comment by Peter-J-Jansen** [comment]
Hi Jürgen,

Thanks for looking into this.

Exactly which "Makefile" and "libtool" files in the SDL-Hercules repo need to be patched this way please ? I only find a "libtool.m4" in directory "autoconf", and several "Makefile.*" files, but none of these without any file type extension. Am I looking in the right places ? Neither "Makefile.am" nor "libtool.m4" match what the patch expects.

Cheers,

Peter

**Comment by Juergen-Git** [comment]
Hi Peter

This really is a meant as a temporary ad'hoc workaround only ;-). So, you do these changes _after_ you've run autogen.sh and ./configure, but before running make. At that point in the process those two files are in the to level directory of the build tree.

Cheers
Jürgen

**Comment by Peter-J-Jansen** [comment]
Thanks for your help Jürgen, it works fine now !
As you predicted, these are the libtool messages : 
```
[...]
  CCLD     hdt3420.la

*** Warning: Linking the shared library hdt3420.la against the
*** static library ../SDL-extpkgs/lib/libcrypto64.a is not portable!

*** Warning: Linking the shared library hdt3420.la against the
*** static library ../SDL-extpkgs/lib/libdecNumber64.a is not portable!

*** Warning: Linking the shared library hdt3420.la against the
*** static library ../SDL-extpkgs/lib/libSoftFloat64.a is not portable!

*** Warning: Linking the shared library hdt3420.la against the
*** static library ../SDL-extpkgs/lib/libtelnet64.a is not portable!
libtool: link: libtool library `hdt3420.la' should begin with `lib'
Try `libtool --help --mode=link' for more information.
  CCLD     vmfplc2
[...]
```
But I presume that's OK then.

Cheers,
Peter

**Comment by Juergen-Git** [comment]
Yes, exactly, these messages are expected and normal with this workaround…

**Comment by Fish-Git** [comment]
> Yes, it should of course read -no-undefined without the blank. I must have messed this up during the many copy&paste tries that didn‘t work...

Fixed!

**Comment by Fish-Git** [comment]
> Yes, exactly, these messages are expected and normal with this workaround…

Thanks, Jürgen!

Now all we have to do is figure out how to make this change work permanently without impacting non-Apple builds. Let me see if I can figure something out...

**Comment by Fish-Git** [comment]
FYI: in case anyone didn't already know this, by clicking the '...' link next to the smiley face at the top of the GitHub comment 
(to the right of the "Member" button) you can select 'Edit' and edit your comment. I do it all the time for not only my own comments, but also for other people's comments too (to fix the formatting to try and make them more readable).

**Comment by Fish-Git** [comment]
FYI: you can also click the "edited" link too, to see what was changed by previous edits.

**Comment by daviddurand** [comment]
I was able to the create a patch that seems to fix the build. I don't understand automake, etc. astonishingly well, but the problem seemed to be that `hdt3420.la` was being built as a hercules module, and the library options were different in that case. I created a new .la that's not a module, and configured it like other non-MOD libraries.

That may compile the same files twice, I'm not sure, but my `Makefile.am` mods seem to create a working makefile after `./autoget.sh`, and the resulting vmfplc2 runs just fine. I expect that this won't mess up Linux builds, but I have no idea about Windows.

I hope this is helpful, and I could make a pull request if you want, but here's the patch:
[patch.txt](https://github.com/SDL-Hercules-390/hyperion/files/3000709/patch.txt)
```diff
diff --git a/Makefile.am b/Makefile.am
index 261ec895..823c3e9d 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -177,6 +177,7 @@ HERCLIBS2 =   \
   libhercu.la \
   libherct.la \
   libhercd.la \
+  libhdt3420_not_mod.la \
   libherc.la
 
 HERCMODS =      \
@@ -371,6 +372,18 @@ libhercd_la_LIBADD = \
   libhercs.la        \
   libhercu.la
 
+
+libhdt3420_not_mod_la_SOURCES  = tapedev.c tapeccws.c awstape.c faketape.c hettape.c omatape.c scsitape.c scsiutil.c
+libhdt3420_not_mod_la_LDFLAGS  =  $(LIB_LD_FLAGS)
+libhdt3420_not_mod_la_LIBADD   = \
+  $(ldadd) \
+  libherct.la \
+  libhercs.la \
+  libhercu.la \
+  libherc.la \
+  libherct.la
+
+
 #--------------------------------------------
 # Pure Utility functions
 #--------------------------------------------
@@ -683,8 +696,8 @@ dmap2hrc_SOURCES   = dmap2hrc.c
 dmap2hrc_LDADD     = $(tools_ADDLIBS)
 dmap2hrc_LDFLAGS   = $(tools_LD_FLAGS)
 
-vmfplc2_SOURCES    = vmfplc2.c
-vmfplc2_LDADD      = $(tools_ADDLIBS) hdt3420.la
+vmfplc2_SOURCES    = vmfplc2.c 
+vmfplc2_LDADD      = $(tools_ADDLIBS) libhdt3420_not_mod.la
 vmfplc2_LDFLAGS    = $(tools_LD_FLAGS)
 
 #------------------------------------------------------------------------------
```

**Comment by Fish-Git** [comment]
@daviddurand

_**THANK YOU**, David!!_ &nbsp; `:D`

Fix commited (a98801b823758e3c8d9689498bea2c084e643796) -- _with credit given to to you!_

**Comment by Peter-J-Jansen** [comment]
David,

I thought I'd also let you know that your fix which Fish committed also worked for me, both on MacOS 10.13 and 10.14.

Thanks a lot !

Cheers,

Peter
